backbone/src/main/java/org/researchstack/backbone/ui/ConsentTaskActivity.kt

```diff
@@ -0,0 +1,219 @@
+    }
+
+    override fun close(completed: Boolean) {
+        // you can also set title / message
+        val dialog = AlertDialog.Builder(this).setCancelable(false)
+                .setTitle(LocaleUtils.getLocalizedString(this, R.string.rsb_saving_consent))
+                .setMessage(LocaleUtils.getLocalizedString(this, R.string.rsb_please_wait)).create()
+
+        dialog.show()
+
+        val consentAssetsFolder = intent.getStringExtra(EXTRA_ASSETS_FOLDER)
+        val role = LocaleUtils.getLocalizedString(this, R.string.rsb_consent_role)
+        val df = DateFormat.getDateInstance(DateFormat.MEDIUM,
+                LocaleUtils.getLocaleFromString(LocaleUtils.getPreferredLocale(this)))
+        consentHtml += getSignatureHtmlContent(getFormalName(firstName, lastName), role, signatureBase64,
+                df.format(Date()))
+
+        PDFWriteExposer().printPdfFile(this, getCurrentTaskId(), consentHtml!!, consentAssetsFolder) {
+            dialog.dismiss()
+            super.close(completed)
+        }
+    }
+
+    private fun getSignatureHtmlContent(completeName: String?, role: String?, signatureB64: String?,
+                                        signatureDate: String?): String {
+        val body = StringBuffer()
+
+        val hr = "<hr align='left' width='100%' style='height:1px; border:none; color:#000; background-color:#000; margin-top: 5px; margin-bottom: 0px;'/>"
+        val td = "<td style='vertical-align: bottom;" + "width: 33%;" + "padding-right: 10px;" + "padding-left: 10px;'>"
+
+        val signatureElementWrapper = "<div class='sigbox'><div class='inbox'>%s</div></div>"
+        val imageTag: String
+
+        val table = ArrayList<Array<String>>()
+
+        if (role == null) {
+            throw RuntimeException("Consent role cannot be empty")
+        }
+
+        // Signature
+        if (completeName != null) {
+            val columnName = getString(R.string.rsb_consent_doc_line_printed_name, role)
+            val columnValue = String.format(signatureElementWrapper, completeName)
+            val column = arrayOf(columnValue, hr, columnName)
+            table.add(column)
+        }
+
+        if (signatureB64 != null) {
+            imageTag = "<img width='100%' alt='star' style='max-height:100px;max-width:200px;height:auto;width:auto;'" +
+                    " src='data:image/png;base64,$signatureB64'/>"
+            val columnName = getString(R.string.rsb_consent_doc_line_signature, role)
+            val columnValue = String.format(signatureElementWrapper, imageTag)
+            val column = arrayOf(columnValue, hr, columnName)
+            table.add(column)
+        }
+
+        if (table.size > 0) {
+            val columnName = getString(R.string.rsb_consent_doc_line_date)
+            val columnValue = String.format(signatureElementWrapper, signatureDate!!)
+            val column = arrayOf(columnValue, hr, columnName)
+            table.add(column)
+        }
+
+        if (table.size > 0) {
```

Comment:
Same as above also.. why are there two similar checks here? :D See the one in line 130 and this one... they are both checking the same, can this be made better?

Reply:
Yea this is very strange! the first check doesn't remove anything from the list so there shouldn't be a second check!